contracts/src/token/erc721/extensions/consecutive.rs

```diff
@@ -0,0 +1,1079 @@
+        }
+    }
+
+    /// Mint a batch of tokens of length `batch_size` for `to`. Returns the
+    /// token id of the first token minted in the batch; if `batchSize` is
+    /// 0, returns the number of consecutive ids minted so far.
+    ///
+    /// Requirements:
+    ///
+    /// - `batchSize` must not be greater than [`MAX_BATCH_SIZE`].
+    /// - The function is called in the constructor of the contract (directly or
+    ///   indirectly).
+    ///
+    /// CAUTION: Does not emit a `Transfer` event. This is ERC-721 compliant as
+    /// long as it is done inside of the constructor, which is enforced by
+    /// this function.
+    ///
+    /// CAUTION: Does not invoke `onERC721Received` on the receiver.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - Write access to the contract's state.
+    /// * `token_id` - Token id as a number.
+    ///
+    /// # Errors
+    ///
+    /// If to is [`Address::ZERO`] error [`rc721::Error::InvalidReceiver`] is
+    /// returned.
+    /// If batch size exceeds [`MAX_BATCH_SIZE`] error
+    /// [`Error::ERC721ExceededMaxBatchMint`] is returned.
+    ///
+    /// # Events
+    ///
+    /// Emits a [`ConsecutiveTransfer`] event.
+    pub fn _mint_consecutive(
+        &mut self,
+        to: Address,
+        batch_size: U96,
+    ) -> Result<U96, Error> {
+        let next = self._next_consecutive_id();
+
+        if batch_size > U96::ZERO {
+            if self._initialized.get() {
+                return Err(ERC721ForbiddenBatchMint {}.into());
+            }
+
+            if to.is_zero() {
+                return Err(erc721::Error::InvalidReceiver(
+                    ERC721InvalidReceiver { receiver: Address::ZERO },
+                )
+                .into());
+            }
+
+            if batch_size > MAX_BATCH_SIZE.to() {
+                return Err(ERC721ExceededMaxBatchMint {
+                    batchSize: U256::from(batch_size),
+                    maxBatch: U256::from(MAX_BATCH_SIZE),
+                }
+                .into());
+            }
+
+            let last = next + batch_size - uint!(1_U96);
+            self._sequential_ownership
+                .push(last, U160::from_be_bytes(to.into_array()))?;
+
+            self.erc721._increase_balance(to, U128::from(batch_size));
+            evm::log(ConsecutiveTransfer {
+                fromTokenId: next.to::<U256>(),
+                toTokenId: last.to::<U256>(),
+                fromAddress: Address::ZERO,
+                toAddress: to,
+            });
+        };
+        Ok(next)
+    }
+
+    /// Should be called to restrict consecutive mint after.
+    /// After this function being called, every call to
+    /// [`Self::_mint_consecutive`] will fail.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - Write access to the contract's state.
+    pub fn _stop_mint_consecutive(&mut self) {
+        self._initialized.set(true);
+    }
```

Comment:
nit: Let's inline this function at the call site. It's not called anywhere else.

Reply:
> This function is assumed to be called a the end user's side

Note that `user` here is a developer, not a user of the deployed contract.

This is an artifact of having `_initialized` as we discussed previously -- functions called from a constructor in Solidity won't be included in the deployed contract, that's why `mint_consecutive` *cannot* be called after the contract is deployed, and why it's a requirement in Solidity, see [the comments](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/19a657bef8354f2a655900654955739b70dfbde9/contracts/token/ERC721/extensions/ERC721Consecutive.sol#L95).

---

This actually surfaces a different issue -- We expect consumers to implement a complex `init` function. I think we should instead implement `init` ourselves, and disallow calling by checking if `initialized == true` at the start of the function. I see we have this function in the example:

```rust
pub fn init(
    &mut self,
    receivers: Vec<Address>,
    batches: Vec<U256>,
) -> Result<(), Error> {
    let len = batches.len();
    for i in 0..len {
        let receiver = receivers[i];
        let batch = batches[i];
        let _ = self
            .erc721_consecutive
            ._mint_consecutive(receiver, U96::from(batch))?;
    }
    self.erc721_consecutive._stop_mint_consecutive();
    Ok(())
}
```

Shouldn't this live in our contract? And then users just call `self.erc721_consecutive.init()`, which in turn sets `initialized`?

---

I also see we have a `constructor.sol` defined for this example, but we don't really need it since we are expecting users to call `init()`?

I think we either choose to have a constructor, which I think is what we should do, or instead have the `init()` mechanism.

Reply:
Yeah, I gave this a bit more thought and I am more convinced that the way to go is to just use a Solidity constructor. It doesn't matter if they are too big.

This is better because:

- It's consistent with the rest of the library.
- We would want it to get audited.
- A user would probably not change much in the constructor of this particular contract.
- It's more secure. (Less vectors vs initializer)

Let's give it a shot 🙏

Reply:
@qalisander wouldn't this be more work? Since we are just adding the contract I think we should aim to do it with a constructor(for consistency and all the other reasons that @alexfertel detailed above). Could you give it a shot so we can see how feasible it is and the level of effort? Thanks!

Reply:
@ggonzalez94 I'm just offering more granular way to make changes. Since I've already tried an approach with solidity constructor and it wasn't compiling that simple. It can take a bit more time. But yeah I'll definitely revisit it in the other branch